crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -0,0 +1,289 @@
+use std::cmp::Ordering;
+
+use ruff_db::files::{File, FilePath};
+
+use crate::db::Db;
+use crate::types::Type;
+
+use super::{
+    class_base::ClassBase, Class, ClassLiteralType, DynamicType, InstanceType, KnownClass,
+    KnownInstanceType, TodoType,
+};
+
+/// Return an [`Ordering`] that describes the canonical order in which two types should appear
+/// in an [`crate::types::IntersectionType`] or a [`crate::types::UnionType`] in order for them
+/// to be compared for equivalence.
+///
+/// Two unions with equal sets of elements will only compare equal if they have their element sets
+/// ordered the same way.
+///
+/// ## Why not just implement `Ord` on `Type`?
+///
+/// It would be fairly easy to slap `#[derive(PartialOrd, Ord)]` on `Type`. However, this would
+/// order types according to their Salsa ID. While this would mean that types would always be
+/// consistently ordered in any single run of red-knot, the order in which they would appear
+/// might vary between different runs of red-knot. The ordering we create here is not user-facing;
+/// however, it will be useful in the future to know where, for example, any `Protocol` elements
+/// lie in the union's element list (if there are any `Protocol` elements in the union's element
+/// list).
+///
+/// Moreover, it doesn't really "make sense" for `Type` to implement `Ord` in terms of the
+/// semantics. There are many different ways in which you could plausibly sort a list of types;
+/// this is only one (somewhat arbitrary, at times) possible ordering.
+pub(super) fn order_union_elements<'db>(
+    db: &'db dyn Db,
+    left: &Type<'db>,
+    right: &Type<'db>,
+) -> Ordering {
+    if left == right {
+        return Ordering::Equal;
+    }
+
+    match (left, right) {
+        (Type::Never, _) => Ordering::Less,
+        (_, Type::Never) => Ordering::Greater,
+
+        (Type::LiteralString, _) => Ordering::Less,
+        (_, Type::LiteralString) => Ordering::Greater,
+
+        (Type::BooleanLiteral(left), Type::BooleanLiteral(right)) => left.cmp(right),
+        (Type::BooleanLiteral(_), _) => Ordering::Less,
+        (_, Type::BooleanLiteral(_)) => Ordering::Greater,
+
+        (Type::IntLiteral(left), Type::IntLiteral(right)) => left.cmp(right),
+        (Type::IntLiteral(_), _) => Ordering::Less,
+        (_, Type::IntLiteral(_)) => Ordering::Greater,
+
+        (Type::StringLiteral(left), Type::StringLiteral(right)) => {
+            left.value(db).cmp(right.value(db))
+        }
```

Comment:
now that the ordering is no longer user-facing, I'm not sure that Salsa db lookups like this in the ordering algorithm make sense anymore. I think it's still useful to make sure that all the `StringLiteral` elements are next to each other in the elements list, but I don't think it matters anymore whether `Literal["bar"]` comes before `Literal["foo"]`. So maybe I should just sort by the Salsa IDs here, which would be cheaper?

Reply:
That makes sense to me

Reply:
I implemented this in https://github.com/astral-sh/ruff/pull/15516/commits/b022cf78ea8b42c9129cb8e1b5f39cdc1dd214d4

Reply:
The version where we first check to see whether it is already sorted before attempting to sort it does (I think) seem to be doing _slightly_ worse on the benchmarks? But it's all within the noise range; I might be seeing signal in random variation here.